src/test/java/bowlingexercise3/BowlingTest.java

```diff
@@ -61,22 +81,339 @@ public void testBowlingSpares(String turns, int score) {
     @CsvSource(value = {
 
             //Strike tests frames 1-9
-            "X, 10",
-            "X|--, 10",
-            "X|11|, 14",
-            "X|5/|1-, 32",
-            "X|X|X|22, 70",
-            "X|43|-8|X|53|-4|45|-7|X, 88",
-            "X|43|X|X|X|X|45|X|X, 166",
-            "X|9/|X|X|52|X|45|X|X, 147",
-            "-/|17|X|-/|5/|X|X|9/|X, 153",
-            "-/|X|X|-/|-/|X|-/|7/|-/, 147",
-            "X|X|X|X|X|X|X|X|X, 240"})
-
-    public void testBowlingStrikes(String turns, int score) {
+            "X|--|--|--|--|--|--|--|--|--||, 10",
+            "X|X|--|--|--|--|--|--|--|--||, 30",
+            "X|11|--|--|--|--|--|--|--|--||, 14",
+            "X|5/|1-|--|--|--|--|--|--|--||, 32",
+            "X|X|X|22|--|--|--|--|--|--||, 70",
+            "X|43|-8|X|53|-4|45|-7|X|--||, 88",
+            "X|43|X|X|X|X|45|X|X|--||, 166",
+            "X|9/|X|X|52|X|45|X|X|--||, 147",
+            "-/|17|X|-/|5/|X|X|9/|X|--||, 153",
+            "-/|X|X|-/|-/|X|-/|7/|-/|--||, 147",
+            "X|X|X|X|X|X|X|X|X|--||, 240"
+    })
+
+    public void testBowlingStrikesFramesOneToNine(String turns, int score) {
+        Bowling game = new Bowling();
+        game.turns(turns);
+        assertEquals(score, game.score());
+    }
+
+    @ParameterizedTest
+    @CsvSource(value = {
+
+            //Testing strikes and spares for all 10 frames including bonus ball
+            //Spare test with and without bonus ball
+            "-/|-/|-/|-/|-/|-/|-/|-/|-/|-/||8, 108",
+            "-/|-/|-/|-/|-/|-/|-/|-/|-/|-/||-, 100",
+            //Strike test with and without bonus ball
+            "X|X|X|X|X|X|X|X|X|X||X-, 290",
+            "X|X|X|X|X|X|X|X|X|X||XX, 300",
+            //Strike, spare and random number test
+            "-/|X|-/|X|-/|X|-/|X|-/|5/||X, 195",
+            "X|-/|X|-/|X|-/|X|-/|X|-/||X, 200",
+            "X|-/|X|45|8/|X|X|7/|X|9/||X, 195",
+
+    })
+
+    public void testBowlingStrikeSpareScoreBonusFrames(String turns, int score) {
         Bowling game = new Bowling();
         game.turns(turns);
         assertEquals(score, game.score());
     }
 
-}
\ No newline at end of file
+    @ParameterizedTest
+    @CsvSource(value = {
+
+            //Testing boundary conditions for spare bonus
+
+            "--|--|--|--|--|--|--|--|--|-/||-",
+            "--|--|--|--|--|--|--|--|--|-/||1",
+            "--|--|--|--|--|--|--|--|--|-/||2",
+            "--|--|--|--|--|--|--|--|--|-/||3",
+            "--|--|--|--|--|--|--|--|--|-/||4",
+            "--|--|--|--|--|--|--|--|--|-/||6",
+            "--|--|--|--|--|--|--|--|--|-/||7",
+            "--|--|--|--|--|--|--|--|--|-/||8",
+            "--|--|--|--|--|--|--|--|--|-/||9",
+            "--|--|--|--|--|--|--|--|--|-/||X",
+            "--|--|--|--|--|--|--|--|--|-/||5",
+    })
+
+    public void boundaryConditionsSpareBonusBallTest(String turns) {
```

Comment:
These are input validation tests for 10th frame containing a spare?

Reply:
These are input validations for a bonus on the 10th frame which is a spare.

Reply:
Would it be called something like assertValidInputForSpareBonusBalls()? What are the boundary conditions?

Reply:
Addressed and renamed appropriately